pkg/limayaml/validate.go

```diff
@@ -160,28 +162,56 @@ func Validate(y LimaYAML) error {
 		// Not validating that the various GuestPortRanges and HostPortRanges are not overlapping. Rules will be
 		// processed sequentially and the first matching rule for a guest port determines forwarding behavior.
 	}
+
+	if err := validateNetwork(y.Network); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func validateNetwork(yNetwork Network) error {
 	networkName := make(map[string]int)
-	for i, vde := range y.Network.VDE {
+	for i, vde := range yNetwork.VDE {
 		field := fmt.Sprintf("network.vde[%d]", i)
 		if vde.URL == "" {
 			return errors.Errorf("field `%s.url` must not be empty", field)
 		}
 		// The field is called VDE.URL in anticipation of QEMU upgrading VDE2 to VDEplug4,
-		// but right now the only valid value is a path to the vde_switch socket directory.
-		fi, err := os.Stat(vde.URL)
-		if err != nil {
-			return errors.Wrapf(err, "field `%s.url` %q failed stat", field, vde.URL)
-		}
-		if !fi.IsDir() {
-			return errors.Wrapf(err, "field `%s.url` %q is not a directory", field, vde.URL)
-		}
-		ctlSocket := filepath.Join(vde.URL, "ctl")
-		fi, err = os.Stat(ctlSocket)
-		if err != nil {
-			return errors.Wrapf(err, "field `%s.url` control socket %q failed stat", field, ctlSocket)
-		}
-		if fi.Mode()&os.ModeSocket == 0 {
-			return errors.Errorf("field `%s.url` file %q is not a UNIX socket", field, ctlSocket)
+		// but right now the only valid value on macOS is a path to the vde_switch socket directory,
+		// optionally with vde:// prefix.
+		if !strings.Contains(vde.URL, "://") || strings.HasPrefix(vde.URL, "vde://") {
+			vdeSwitch := strings.TrimPrefix(vde.URL, "vde://")
+			fi, err := os.Stat(vdeSwitch)
+			if err != nil {
+				return errors.Wrapf(err, "field `%s.url` %q failed stat", field, vdeSwitch)
+			}
+			if fi.IsDir() {
+				/* Switch mode (vdeSwitch is dir, port != 65535) */
+				ctlSocket := filepath.Join(vdeSwitch, "ctl")
+				fi, err = os.Stat(ctlSocket)
+				if err != nil {
+					return errors.Wrapf(err, "field `%s.url` control socket %q failed stat", field, ctlSocket)
+				}
+				if fi.Mode()&os.ModeSocket == 0 {
+					return errors.Errorf("field `%s.url` file %q is not a UNIX socket", field, ctlSocket)
+				}
+				if vde.Port == 65535 {
+					return errors.Errorf("field `%s.url` points to a non-PTP switch, so the port number must not be 65535", field)
+				}
+			} else {
+				/* PTP mode (vdeSwitch is socket, port == 65535) */
+				if fi.Mode()&os.ModeSocket == 0 {
+					return errors.Errorf("field `%s.url` %q is not a directory nor a UNIX socket", field, vdeSwitch)
+				}
+				if vde.Port != 65535 {
+					return errors.Errorf("field `%s.url` points to a PTP (switchless) socket %q, so the port number has to be 65535 (got %d)",
+						field, vdeSwitch, vde.Port)
+				}
+			}
+		} else if runtime.GOOS != "linux" {
+			logrus.Warnf("field %s.url is unlikely to work for %s (unless you managed to port over libvdeplug4 to %s by yourself)",
```

Comment:
Put `%s.url` in backquotes.

The quoted text feels slightly snarky to me. Maybe "(unless libvdeplug4 has been ported to %s and is installed)" is more neutral?

Reply:
updated